rust/src/operations/delete.rs

```diff
@@ -0,0 +1,1074 @@
+        DataType::Int8 => append_builder!(builder, value, Int8Builder, i8),
+        DataType::Int16 => append_builder!(builder, value, Int16Builder, i16),
+        DataType::Int32 => append_builder!(builder, value, Int32Builder, i32),
+        DataType::Int64 => append_builder!(builder, value, Int64Builder, i64),
+        DataType::UInt8 => append_builder!(builder, value, UInt8Builder, u8),
+        DataType::UInt16 => append_builder!(builder, value, UInt16Builder, u16),
+        DataType::UInt32 => append_builder!(builder, value, UInt32Builder, u32),
+        DataType::UInt64 => append_builder!(builder, value, UInt64Builder, u64),
+        DataType::Float32 => append_builder!(builder, value, Float32Builder, f32),
+        DataType::Float64 => append_builder!(builder, value, Float64Builder, f64),
+        DataType::Utf8 => {
+            let array = builder
+                .as_any_mut()
+                .downcast_mut::<StringBuilder>()
+                .unwrap();
+            array.append_option(value);
+        }
+        _ => {
+            return Err(DeltaTableError::Generic(format!(
+                "Unable to convert value to arrow type {}",
+                t
+            )))
+        }
+    }
+
+    Ok(())
+}
+
+fn new_array_builder(t: DataType, capacity: usize) -> DeltaResult<Box<dyn ArrayBuilder>> {
+    Ok(match t {
+        DataType::Boolean => Box::new(BooleanBuilder::with_capacity(capacity)),
+        DataType::Int8 => Box::new(Int8Builder::with_capacity(capacity)),
+        DataType::Int16 => Box::new(Int16Builder::with_capacity(capacity)),
+        DataType::Int32 => Box::new(Int32Builder::with_capacity(capacity)),
+        DataType::Int64 => Box::new(Int64Builder::with_capacity(capacity)),
+        DataType::UInt8 => Box::new(UInt8Builder::with_capacity(capacity)),
+        DataType::UInt16 => Box::new(UInt16Builder::with_capacity(capacity)),
+        DataType::UInt32 => Box::new(UInt32Builder::with_capacity(capacity)),
+        DataType::UInt64 => Box::new(UInt64Builder::with_capacity(capacity)),
+        DataType::Float32 => Box::new(Float32Builder::with_capacity(capacity)),
+        DataType::Float64 => Box::new(Float64Builder::with_capacity(capacity)),
+        DataType::Utf8 => Box::new(StringBuilder::with_capacity(capacity, 1024)),
+        _ => {
+            return Err(DeltaTableError::Generic(format!(
+                "Unable to create builder for arrow type {}",
+                t
+            )))
+        }
+    })
+}
+
+fn get_inner_type(t: DataType) -> DeltaResult<DataType> {
+    match t {
+        DataType::Dictionary(_, value) => Ok(value.as_ref().to_owned()),
+        _ => Err(DeltaTableError::Generic(format!(
+            "Unable to create builder for arrow type {}",
+            t
+        ))),
+    }
+}
+
+// Create a record batch that contains the partition columns plus the path of the file
+fn create_partition_record_batch(
```

Comment:
Maybe you can re-use this function for now:

https://github.com/delta-io/delta-rs/blob/e046a774b608ac7df5114c014a1c22dbb60da38f/rust/src/table_state_arrow.rs#L51

It handles the different data types for partition values. It has a little more overhead since it also parses out the statistics, but I think that's fine for now. Later on, I expect we'll replace this with expression simplification which will let us use statistics and remove redundant parts of the predicate:

https://github.com/apache/arrow-datafusion/issues/6171